Migrate docs from mdbook to docusaurus

Keinberger opened the pull request.

Keinberger commented on docs/src/web-client/api/classes/Account.md:
Thanks for pointing that out.

To understand better how the current docs are structured, would you please mind clarifying the following:
- Which parts of the docs are generated by typedoc, which are not? (does it only apply to the api folder?) Do you largely just rely on this workflow to generate the API docs?
- Are the generated API docs meant to appear on the public docs site, or are they internal-only (they are currently not exported)? Do we intend to publish them in the future? If not, I think it makes sense to decouple these md files from the Docusaurus docs.
- If we do intend on including those files in the Miden Book: Since you already know how this is set up, would you mind pointing out how the current set up can be modified to accommodate for the migration? Would be great if you could give me more context so that I can amend the Makefile or you can also just directly go ahead and make a commit to fix that.

Would also be great if you could answer my original question in the PR description (Questions & Clarifications section).

igamigo replied:
> Which parts of the docs are generated by typedoc, which are not? (does it only apply to the api folder?) Do you largely just rely on this workflow to generate the API docs?

The webclient SDK reference docs are generated by `typedoc` and are output in `./docs/src/web-client/api` (see the [`typedoc.json`](https://github.com/0xMiden/miden-client/blob/next/crates/web-client/typedoc.json) file for a bit more context). It generates docs for the public web client API, listing functions and parameters. It also flows documentation from the Rust side of the bindings.

> Are the generated API docs meant to appear on the public docs site, or are they internal-only (they are currently not exported)? Do we intend to publish them in the future? If not, I think it makes sense to decouple these md files from the Docusaurus docs.

Honestly not sure, that would depend on how we want to structure docs in general. I think @dagarcia7 added this flow months ago but I'm not sure where they are (or should) be advertised. 

> If we do intend on including those files in the Miden Book: Since you already know how this is set up, would you mind pointing out how the current set up can be modified to accommodate for the migration? Would be great if you could give me more context so that I can amend the Makefile or you can also just directly go ahead and make a commit to fix that.

I'm not too familiar with typedoc to know whether something like adapting the API reference generation to docusaurus is easily doable (again, maybe @dagarcia7 knows). If you prefer, we can leave it out for now and we can figure out how it can be adapted as a follow-up. The [`typedoc.json`](https://github.com/0xMiden/miden-client/blob/next/crates/web-client/typedoc.json) file contains some settings, but they don't seem super useful for this. Seems also like there is a [typedoc plugin](https://github.com/typedoc2md/typedoc-plugin-markdown) for docusaurus, but unsure how it could work (probably worth investigating though).

Keinberger commented on docs/src/rust-client/design.md:
```diff
@@ -1,3 +1,8 @@
+---
```
Will fix!

Keinberger commented on docs/src/web-client/examples.md:
No, not quite. This page corresponds to the existing "index" page on the Rust client docs. See here: https://0xmiden.github.io/miden-docs/imported/miden-client/src/web-client/examples/index.html

I pointed this page out in the PR description up above as part of the "Questions" because (given its weird name and positioning) I think this page may not be named correctly or may not be intended to be included where it is right now.

@igamigo Your input on this would be highly appreciated!

dagarcia7 replied:
Completely agree with everything here. We should revisit these examples (again, maybe in a follow up issue) and make sure they're still relevant and working. And happy to revisit the structure and the overall documentation as well.

Keinberger replied:
Okey, I'll change the title of the page to "Examples" for now. We can rework the docs content in other PRs. I think that's fine, do you agree?

igamigo replied:
Sounds good to me!

Keinberger commented:
@igamigo Thanks for approving, but I would still like your take on the following before merging:
- The "Questions & Clarifications" section in the PR description
-  https://github.com/0xMiden/miden-client/pull/1338#discussion_r2390081509 (to know if we want to include the API docs in Docusaurus or not)
- https://github.com/0xMiden/miden-client/pull/1338#discussion_r2390669966 (to know if this file is correctly presented right now)

igamigo commented on docs/src/rust-client/design.md:
```diff
@@ -6,9 +11,10 @@ The Miden client has the following architectural components:
-> [!Tip]
-> - The RPC client and the store are Rust traits.
-> - This allow developers and users to easily customize their implementations.
+:::tip
+- The RPC client and the store are Rust traits.
+- This allow developers and users to easily customize their implementations.
+:::
```
I think this one is not rendering correctly for some reason:

<img width="754" height="155" alt="Image" src="https://github.com/user-attachments/assets/9cbe3e32-36cb-40ee-96fa-cc0a0caee9ce" />

Others look like this:

<img width="1018" height="104" alt="Image" src="https://github.com/user-attachments/assets/560fe5c6-a36d-49fc-bdc2-ea632ab96a3f" />

Keinberger replied:
Thanks for pointing that out, fixed!

igamigo commented:
> * The "Questions & Clarifications" section in the PR description
> * [Migrate docs from mdbook to docusaurus #1338 (comment)](https://github.com/0xMiden/miden-client/pull/1338#discussion_r2390669966) (to know if this file is correctly presented right now)

Mainly meant to respond this with [this comment](https://github.com/0xMiden/miden-client/pull/1338#discussion_r2391940644) and [this other comment](https://github.com/0xMiden/miden-client/pull/1338#discussion_r2391920357). In short, I would include it but change where it's placed in the section hierarchy and name it "Examples" or something similar (page title is fine). But in general, it seems like the whole web-client subsection should be reworked.

> * [Migrate docs from mdbook to docusaurus #1338 (comment)](https://github.com/0xMiden/miden-client/pull/1338#discussion_r2390081509) (to know if we want to include the API docs in Docusaurus or not)

Also meant to be commented on [here](https://github.com/0xMiden/miden-client/pull/1338#discussion_r2391900648). I am not sure as this depends on how we want the overall structure to look. It also depends on whether this is already published somewhere else or if we can publish it anywhere else. Also, because these are generated automatically from exported types,  we'd have to investigate a plugin or a way to make typedoc and docusaurus play well together.

Keinberger commented:
> Also meant to be commented on https://github.com/0xMiden/miden-client/pull/1338#discussion_r2391900648. I am not sure as this depends on how we want the overall structure to look. It also depends on whether this is already published somewhere else or if we can publish it anywhere else. Also, because these are generated automatically from exported types, we'd have to investigate a plugin or a way to make typedoc and docusaurus play well together.

I think it would make sense to separately host the typedoc stuff, and then link it from the main docs. What do you think?

igamigo commented:
> > Also meant to be commented on [#1338 (comment)](https://github.com/0xMiden/miden-client/pull/1338#discussion_r2391900648). I am not sure as this depends on how we want the overall structure to look. It also depends on whether this is already published somewhere else or if we can publish it anywhere else. Also, because these are generated automatically from exported types, we'd have to investigate a plugin or a way to make typedoc and docusaurus play well together.
> 
> I think it would make sense to separately host the typedoc stuff, and then link it from the main docs. What do you think?

Yes, agreed as well!

Keinberger commented:
Hey @igamigo ,

Based on discussions I had with Philipp Gackstatter and Bobbin on the Docusaurus PR on the `miden-base` repository, I also included some more changes on this branch:

- `_category_.json` files are now in YAML format. That way we can add a comment in each file to explain the position parameter
- Added amendments to the Makefile, to add logic for serving the docs locally
- Added a small section in the README that explains the docs architecture (NPM dependencies and absorption into main docs page)
- Added a new build-docs workflow to check md syntax for the exported docs
- `trigger-deploy.yml` workflow
    - Added a specific SHA ref for safety (since it uses a personal access token)
    - Amended the workflow to only trigger on changes to the `docs/external` directory

Would be great if you could give a final sign-off on that!

igamigo commented on .github/workflows/build-docs.yml:
```diff
@@ -0,0 +1,46 @@
+
+      - name: Build documentation
+        working-directory: ./docs/external
+        run: npm run build:dev
```
nit: Usually I'd try to run `make` commands here (eg `make build-docs`), but this is simple enough so feel free to disregard

igamigo commented on Makefile:
```diff
@@ -77,9 +77,9 @@ doc: ## Generate & check rust documentation. You'll need `jq` in order for this
-book: ## Builds the book & serves documentation site
-	mdbook serve --open docs
+.PHONY: serve-docs
+serve-docs: ## Serves the docs
```
With these new changes, is there anything we need to add to the `check-tools`/`install-tools` commands?

Keinberger replied:
No, `npm` is already included there (bc of yarn) and that's the only thing that is needed to run the docs.

Keinberger commented on .github/workflows/trigger-deploy-docs.yml:
```diff
@@ -0,0 +1,23 @@
+
+on:
+  push:
+    branches: [next]
```
Changes to `main` indicate a new "release". "Release" snapshots are manually taken inside of the miden-docs repository using the `cut-versions.yml` workflow.

Keinberger commented on README.md:
```diff
@@ -51,6 +51,12 @@ The crate also comes with one feature flag that is used exclusively on tests:
 
+## Docs
+
+The documentation in the `docs/external` folder is built using Docusaurus and is automatically absorbed into the main [miden-docs](https://github.com/0xMiden/miden-docs) repository for the main documentation website. Changes to the `next` branch trigger an automated deployment workflow. The docs folder requires npm packages to be installed before building.
```
I think explaining how the docs are integrated is actually very important, because it is otherwise not apparent that the docs are imported somewhere else, and actually used inside of our main Miden Docs. This is an important detail, because it impacts the way we structure, and also write the docs. I definitely think we need that to be documented in a clear way so that it is obvious to anyone working on the repo.

What do you think?

evanmarshall commented:
Can we update this link as part of this PR? Currently the link is broken in npmjs's website: https://github.com/0xMiden/miden-client/pull/1372/files#diff-3e75433bb9dbb3c274b64f09230d8ba407b69e56c07ead5964a4867173bbb3faL16
